File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn question? [input]
  "INPUT is determined to be a question."
  (str/ends-with? input "?"))

(defn yelling? [input]
  "INPUT is determined to be yelling."
  (cond
    (not-any? #(Character/isLetter %) input) false
    :else (= input (str/upper-case input))))

(defn yelling-question? [input]
  "INPUT is determined to be yelling a question."
  ((every-pred yelling? question?) input))

(defn silent? [input]
  "INPUT is determined to be silent."
  (str/blank? input))

(defn response-for [input]
  "Bob's response to INPUT."
  (condp apply [(str/trim input)]
    yelling-question? "Calm down, I know what I'm doing!"
    question? "Sure."
    yelling? "Whoa, chill out!"
    silent? "Fine. Be that way!"
    "Whatever."))

File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      silent?(input) ->
        "Fine. Be that way!"
      asking?(input) and shouting?(input) ->
        "Calm down, I know what I'm doing!"
      asking?(input) ->
        "Sure."
      shouting?(input) ->
        "Whoa, chill out!"
      true ->
        "Whatever."
    end
  end

  defp asking?(statement) do
    statement
    |> String.trim
    |> String.ends_with?("?")
  end

  defp shouting?(statement) do
    statement == String.upcase(statement) and statement != String.downcase(statement)
  end

  defp silent?(statement) do
    "" == statement |> String.trim
  end
end
